core/embed/rust/src/ui/display.rs

```diff
@@ -105,6 +153,345 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+        let mut overlay_color = None;
+
+        if self.area.contains(p) {
+            let mut tot_adv = 0;
+
+            let p_rel = Point::new(p.x - self.area.x0, p.y - self.area.y0);
+
+            for c in self.text.bytes() {
+                if let Some(g) = self.font.get_glyph(c) {
+                    let w = g.width;
+                    let h = g.height;
+                    let b_x = g.bearing_x;
+                    let b_y = g.bearing_y;
+
+                    let char_area = Rect::new(
+                        Point::new(tot_adv + b_x, h - b_y),
+                        Point::new(tot_adv + b_x + w, b_y),
+                    );
+
+                    if char_area.contains(p_rel) {
+                        let p_inner = p_rel - char_area.top_left();
+
+                        let overlay_data = g.get_pixel_data(p_inner);
+
+                        if overlay_data > 0 {
+                            if let Some(u) = underlying {
+                                overlay_color = Some(Color::lerp(
+                                    self.colortable[15],
+                                    u,
+                                    overlay_data as f32 / 15_f32,
+                                ));
+                            } else {
+                                overlay_color = Some(self.colortable[overlay_data as usize]);
+                            }
+                        }
+                        break;
+                    }
+                    tot_adv += g.adv;
+                }
+            }
+        }
+
+        overlay_color
+    }
+}
+
+/// Performs a conversion from `angle` (in degrees) to a vector (`Point`)
+/// (polar to cartesian transformation)
+/// Suitable for cases where we don't care about distance, it is assumed 1000
+///
+/// The implementation could be replaced by (cos(`angle`), sin(`angle`)),
+/// if we allow trigonometric functions.
+/// In the meantime, approximate this with predefined octagon
+fn get_vector(angle: i32) -> Point {
+    //octagon vertices
+    let v = [
+        Point::new(0, 1000),
+        Point::new(707, 707),
+        Point::new(1000, 0),
+        Point::new(707, -707),
+        Point::new(0, -1000),
+        Point::new(-707, -707),
+        Point::new(-1000, 0),
+        Point::new(-707, 707),
+    ];
+
+    match angle % 360 {
+        0..=44 => Point::lerp(v[0], v[1], (angle) as f32 / 45_f32),
+        45..=89 => Point::lerp(v[1], v[2], (angle - 45) as f32 / 45_f32),
+        90..=134 => Point::lerp(v[2], v[3], (angle - 90) as f32 / 45_f32),
+        135..=179 => Point::lerp(v[3], v[4], (angle - 135) as f32 / 45_f32),
+        180..=224 => Point::lerp(v[4], v[5], (angle - 180) as f32 / 45_f32),
+        225..=269 => Point::lerp(v[5], v[6], (angle - 225) as f32 / 45_f32),
+        270..=314 => Point::lerp(v[6], v[7], (angle - 270) as f32 / 45_f32),
+        315..=359 => Point::lerp(v[7], v[0], (angle - 315) as f32 / 45_f32),
+        _ => Point::new(1000, 0),
+    }
```

Comment:
Using your suggestion: https://github.com/trezor/trezor-firmware/commit/e18e0fbdf4ded6fc452cedd84cfe119a5bf53d99

about sin/con: one thing is that we don't satisfy `#[lang = "f32_runtime"]` so we don't have these available (and i don't know how to enable this and what other implications there might be if any). Then we would need to evaluate flash usage and execution time. But if we are gonna do some more animations and graphics, these might come handy, so it would be worth experimenting with at least.

Reply:
Tried it with sin/cos: we don't have these function in rust since they are part of std, which we don't use. Tried using rust crate libm, which worked but it produced about 7kB larger binary. Tried also compiling&exposing C sin/cos functions, which is a bit better - about 4kB more. Didn't check the performance as the size is imo blocker until we really need it.